l10n_it_delivery_note/models/stock_delivery_note.py

```diff
@@ -433,36 +443,61 @@ def action_confirm(self):
                 note.name = sequence.next_by_id()
                 note.sequence_id = sequence
 
-    def _fix_quantities_to_invoice(self, lines):
+    def _check_delivery_notes_before_invoicing(self):
+        for delivery_note_id in self:
+            if not delivery_note_id.sale_ids:
+                raise UserError(
+                    _("%s hasn't sale order!") % delivery_note_id.display_name
+                )
+            if delivery_note_id.invoice_status == "invoiced":
+                raise UserError(
+                    _("%s is already invoiced!") % delivery_note_id.display_name
+                )
+            if delivery_note_id.state == "draft":
+                raise UserError(_("%s is in draft!") % delivery_note_id.display_name)
+            for line in delivery_note_id.line_ids:
+                if line.product_id.invoice_policy == "order":
+                    raise UserError(
+                        _("In %s there is %s with invoicing policy 'order'")
+                        % (delivery_note_id.display_name, line.product_id.name)
+                    )
+
+    def _fix_quantities_to_invoice(self, lines, invoice_method):
         cache = {}
 
         pickings_lines = lines.retrieve_pickings_lines(self.picking_ids)
```

Comment:
self qui è un recordset (da correggere anche in v.14.0)
```suggestion
        pickings_lines = lines.retrieve_pickings_lines(self.mapped("picking_ids"))
```